ducktape/tests/test.py

```diff
@@ -32,17 +33,85 @@ def __init__(self, test_context):
         self.cluster = test_context.session_context.cluster
         self.test_context = test_context
         self.logger = test_context.logger
+        self.services = ServiceRegistry()
+
+        # dict for toggling service log collection on/off
+        self.log_collect = {}
 
     def min_cluster_size(self):
         """
-        Subclasses implement this to provide a helpful heuristic to prevent trying to run a test on a cluster
-        with too few nodes.
+        Provides a helpful heuristic for determining if there are enough nodes in the cluster to run this test.
         """
-        raise NotImplementedError("All tests must implement this method.")
+        return self.services.num_nodes()
+
+    def allocate_nodes(self):
+        self.services.allocate_nodes()
 
-    def service_context(self, num_nodes):
+    def tearDown(self):
+        """Default teardown method which stops and cleans services registered in the ServiceRegistry.
+        Note that there is not a default setUp method. This is because the framework makes no assumptions
+        about when and in what methods the various services are started.
+        """
+        if hasattr(self, 'services') and type(self.services) == ServiceRegistry:
+            self.services.stop_all()
+            self.copy_service_logs()
+            self.services.clean_all()
+
+    def free_nodes(self):
+        self.services.free_all()
+
+    def copy_service_logs(self):
+        """Copy logs from service nodes to the results directory."""
+        for service in self.services.values():
+            if not hasattr(service, 'logs'):
+                self.test_context.logger.debug("Won't collect service logs from %s - no 'logs' attribute." %
+                    service.__class__.__name__)
+                return
+
+            log_dirs = service.logs
+            for node in service.nodes:
+                for log_name in log_dirs.keys():
```

Comment:
The original idea was to provide finer-grained options. The use-case I had in mind was to have the ability to collect the Kafka 'service log' without collecting the kafka 'data log' (or dynamically choose whether to collect the data logs).

Maybe all or none is good enough, I'm not sure.